services/sso/src/main/java/software.amazon.awssdk.services.sso/auth/SsoCredentialsProvider.java

```diff
@@ -0,0 +1,303 @@
+            }
+
+            // Update SSO Credentials with configurations in profile
+            cacheBuilder = CachedSupplier.builder(() -> updateSsoCredentials(profileOptional.get()));
+            if (builder.asyncCredentialUpdateEnabled) {
+                cacheBuilder.prefetchStrategy(new NonBlocking(ASYNC_THREAD_NAME));
+            }
+
+        } catch (RuntimeException e) {
+            loadException = e;
+        }
+
+        if (loadException != null) {
+            this.loadException = loadException;
+            this.credentialCache = null;
+        } else {
+            this.loadException = null;
+            this.credentialCache = cacheBuilder.build();
+        }
+    }
+
+    /**
+     * Update the expiring session SSO credentials by calling SSO. Invoked by {@link CachedSupplier} when the credentials
+     * are close to expiring.
+     */
+    private RefreshResult<SsoCredentialsHolder> updateSsoCredentials(Profile profile) {
+        SsoCredentialsHolder credentials = new SsoCredentialsHolder(
+            (AwsSsoCredentials) new SsoCredentialsUtils(profile, ssoClient).getUpdatedCredentials());
+        Instant acutalTokenExpiration = credentials.getSsoCredentialsExpiration().toInstant();
+
+        return RefreshResult.builder(credentials)
+                            .staleTime(acutalTokenExpiration.minus(staleTime))
+                            .prefetchTime(acutalTokenExpiration.minus(prefetchTime))
+                            .build();
+    }
+
+    /**
+     * The amount of time, relative to session token expiration, that the cached credentials are considered stale and
+     * should no longer be used. All threads will block until the value is updated.
+     */
+    public Duration staleTime() {
+        return staleTime;
+    }
+
+    /**
+     * The amount of time, relative to session token expiration, that the cached credentials are considered close to stale
+     * and should be updated.
+     */
+    public Duration prefetchTime() {
+        return prefetchTime;
+    }
+
+    /**
+     * Get a builder for creating a custom {@link SsoCredentialsProvider}.
+     */
+    public static BuilderImpl builder() {
+        return new BuilderImpl();
+    }
+
+    @Override
+    public AwsCredentials resolveCredentials() {
+        if (loadException != null) {
+            throw loadException;
+        }
+        return credentialCache.get().getSsoCredentials();
+    }
+
+    @Override
+    public void close() {
+        credentialCache.close();
+        ssoClient.close();
+    }
+
+    /**
+     * A builder for creating a custom {@link SsoCredentialsProvider}.
+     */
+    public interface Builder {
+
+        /**
+         * Define the profile file that should be used by this credentials provider. By default, the
+         * {@link ProfileFile#defaultProfileFile()} is used.
+         */
+        Builder profileFile(ProfileFile profileFile);
```

Comment:
Why tie this to the ProfileFile file instead of having the ProfileCredentialProvider create this class like the other providers?